map.go

```diff
@@ -1614,6 +1637,85 @@ func (mi *MapIterator) Next(keyOut, valueOut interface{}) bool {
 	return false
 }
 
+func (mi *MapIterator) nextDrain(keyOut, valueOut interface{}) bool {
+	// Check for maps without a key (e.g., Queue/Stack). In this case, when there is no
+	// more data, ErrKeyNotExist arise, but we gently stop the retrieval with no error.
+	if mi.target.keySize == 0 {
+		mi.err = mi.target.LookupAndDelete(keyOut, valueOut)
+		if errors.Is(mi.err, ErrKeyNotExist) {
+			mi.done = true
+			mi.err = nil
+			return false
+		} else if mi.err != nil {
+			return false
+		}
+		mi.count++
+		return true
+	}
+
+	// Here we allocate only once data for retrieving the next key in the map.
+	if mi.cursor == nil {
+		mi.cursor = make([]byte, mi.target.keySize)
+	}
+
+	// Always retrieve first key in the map. This should ensure that
+	// the whole map is traversed, despite concurrent insertion.
+	// The expected ordering might differ:
+	// - initial keys in map: `a -> b -> c`
+	// - call MapIterator.Next and retrieve key `a`
+	// - insert key `d` in map
+	// - retrieve all the remaining keys `d -> b -> c`
+	mi.err = mi.target.NextKey(nil, mi.cursor)
+	if errors.Is(mi.err, ErrKeyNotExist) {
+		mi.done = true
+		mi.err = nil
+		return false
+	} else if mi.err != nil {
+		mi.err = fmt.Errorf("get next key: %w", mi.err)
+		return false
+	}
+
+	// falling back to sequential Lookup -> Delete in case LookupAndDelete
+	// is not supported (e.g., kernel < 5.14).
+	if mi.fallback {
+		mi.count++
+		mi.err = mi.target.Lookup(mi.cursor, valueOut)
+		if errors.Is(mi.err, ErrKeyNotExist) {
+			// Same as in MapIterator.nextIterate.
+			return mi.nextDrain(keyOut, valueOut)
+		} else if mi.err != nil {
+			mi.err = fmt.Errorf("look up next key: %w", mi.err)
+			return false
+		}
+		mi.err = mi.target.Delete(mi.cursor)
+	} else {
+		// Check if LookupAndDelete is supported and not invalid args, otherwise fallback
+		mi.err = mi.target.LookupAndDelete(mi.cursor, valueOut)
+		if mi.err != nil && errors.Is(mi.err, ErrNotSupported) || errors.Is(mi.err, unix.EINVAL) {
+			mi.fallback = true
+			return mi.nextDrain(keyOut, valueOut)
+		}
+		mi.count++
+	}
+
+	if errors.Is(mi.err, ErrKeyNotExist) {
+		// Same as in MapIterator.nextIterate.
+		return mi.nextDrain(keyOut, valueOut)
```

Comment:
Why does this recurse? To avoid the for loop? I think the for loop is easier to understand here because one just has to look for a break or return statement to figure out how the loop exits. For me that is more difficult with recursive functions.